examples/linear_model/plot_sgd_early_stopping.py

```diff
@@ -0,0 +1,155 @@
+    if n_samples is not None:
+        X, y = X[:n_samples], y[:n_samples]
+    return X, y
+
+
+@ignore_warnings(category=ConvergenceWarning)
+def fit_and_score(estimator, max_iter, X_train, X_test, y_train, y_test):
+    """Fit the estimator on the train set and score it on both sets"""
+    estimator.set_params(max_iter=max_iter)
+
+    start = time.time()
+    estimator.fit(X_train, y_train)
+
+    fit_time = time.time() - start
+    n_iter = estimator.n_iter_
+    train_score = estimator.score(X_train, y_train)
+    test_score = estimator.score(X_test, y_test)
+
+    return fit_time, n_iter, train_score, test_score
+
+
+# Define the estimators to compare
+estimator_dict = {
+    'No stopping criterion':
+    linear_model.SGDClassifier(tol=None, random_state=0),
+    'Training loss':
+    linear_model.SGDClassifier(early_stopping=False, n_iter_no_change=5,
+                               tol=0.01, random_state=0),
+    'Validation score':
+    linear_model.SGDClassifier(early_stopping=True, n_iter_no_change=5,
+                               tol=0.01, random_state=0,
+                               validation_fraction=0.2)
+}
+
+# Load the dataset
+X, y = load_mnist(n_samples=10000)
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.5,
+                                                    random_state=0)
+
+max_iter = 50
+results = []
+for estimator_name, estimator in estimator_dict.items():
+    print(estimator_name + ': ', end='')
+    for this_max_iter in range(1, max_iter):
+        print('.', end='')
+        sys.stdout.flush()
+
+        fit_time, n_iter, train_score, test_score = fit_and_score(
+            estimator, this_max_iter, X_train, X_test, y_train, y_test)
+
+        results.append((estimator_name, this_max_iter, fit_time, n_iter,
+                        train_score, test_score))
+    print('')
+
+results = np.array(results).T
+results_dict = {
+    'Stopping criterion': results[0],
+    'max_iter': results[1],
+    'Fit time (sec)': results[2],
+    'n_iter_': results[3],
+    'Train score': results[4],
+    'Test score': results[5],
+}
+
+# Define what to plot (index, values), i.e. (x_axis, y_axis),
+# for all cases listed in columns.
+columns = 'Stopping criterion'
+plot_list = [
+    ('max_iter', 'Train score'),
+    ('max_iter', 'Test score'),
+    ('max_iter', 'n_iter_'),
+    ('max_iter', 'Fit time (sec)'),
+]
+
+nrows = 2
+ncols = int(np.ceil(len(plot_list) / 2))
+fig, axes = plt.subplots(nrows=nrows, ncols=ncols, figsize=(6 * ncols,
+                                                            4 * nrows))
+axes[0, 0].get_shared_y_axes().join(axes[0, 0], axes[0, 1])
+
+for ax, (index, values) in zip(axes.ravel(), plot_list):
+    for this_column in np.unique(results_dict[columns]):
+        mask = results_dict[columns] == this_column
+        ax.plot(results_dict[index][mask], results_dict[values][mask],
+                label=this_column)
```

Comment:
This code is way too complicated to read (and debug as apparently it fails on recent versions of matplotlib).

I am in favor or loading the raw results into a pandas dataframe and then 

for criteron, goup_df in df.groupby('criterion'):

Reply:
I know but I think pandas-based plots is much easier to read and maintain, so i am +1 in favor of using pandas.